### PR TITLE
[Testcase Refactoring] Add hw_classification and split device-dependent test in test_multi_threaded_pg

### PR DESCRIPTION
**Pull request by twthighd**
## Summary
Split the device-dependent test_bwd_sees_fwd_pg out of the mixed TestCollectivesWithBaseClass into a dedicated ACCELERATOR class instantiated via instantiate_device_type_tests, and label the CPU-only collective classes GENERIC, so the threaded-collective tests are correctly classified under --hw-classification filtering and run on-device via an injected device parameter instead of a module-level device string.

## Changes
- **`test/distributed/test_multi_threaded_pg.py`**:
  - Labeled `TestCollectivesWithWrapper` and `TestCollectivesWithBaseClass` with `hw_classification = HardwareClassification.GENERIC` (device-agnostic threaded-collective logic on CPU tensors).
  - Moved `test_bwd_sees_fwd_pg` (requires an accelerator; asserts that autograd backward runs in the same thread as forward and can still see the process group on a device tensor) out of `TestCollectivesWithBaseClass` into a new parallel `TestCollectivesWithBaseClassDevice` class labeled `hw_classification = HardwareClassification.ACCELERATOR`.
  - Extracted a shared `_TestCollectivesBase(MultiThreadedTestCase)` fixture base holding `world_size` / `setUp` / `tearDown` (the `TORCH_DIST_INIT_BARRIER` env setup plus `_spawn_threads`); both `TestCollectivesWithBaseClass` and `TestCollectivesWithBaseClassDevice` inherit it. The base has no `test_*` methods, so unittest does not collect it and `--hw-classification` filtering does not scan it; no `hw_classification` label is needed on it.
  - `test_bwd_sees_fwd_pg` now takes a `device` parameter, and `TestCollectivesWithBaseClassDevice` is instantiated via `instantiate_device_type_tests(..., except_for=("cpu",), allow_xpu=True)`. The `device` is injected per-variant rather than read from a module-level device string.
  - Removed the module-level `device_type = acc.type if (acc := torch.accelerator.current_accelerator()) else "cpu"` string and the pre-existing `@skip_if_lt_x_gpu(1)` decorator on `test_bwd_sees_fwd_pg` (plus its `skip_if_lt_x_gpu` import). On a CPU-only host the method is no longer collected because `except_for=("cpu",)` excludes the CPU variant at instantiation time; on an accelerator host only the accelerator variant is generated and runs with the injected `device`. This matches the original behavior (the test ran once on the available accelerator) while extending coverage to other accelerators beyond CUDA.

## Motivation
`TestCollectivesWithBaseClass` was a mixed class: 17 CPU threaded-collective methods plus one accelerator-dependent method (`test_bwd_sees_fwd_pg`). Under `--hw-classification` filtering such a mixed class cannot be classified unambiguously, and the accelerator method resolved its device from a module-level string (`device_type = acc.type if ...`) rather than the per-variant `device` parameter injected by `instantiate_device_type_tests` — so non-default device variants would use the wrong device. Splitting the class by hardware dependency and instantiating the ACCELERATOR class with an injected `device` parameter makes the classification unambiguous and aligns with the TEST_LINTER direction (PR #190173): instantiated, `device` param on every `test_*`, no `only_for`.

The `@skip_if_lt_x_gpu(1)` → `except_for=("cpu",)` change also fixes a behavior divergence surfaced in review: with `except_for` removed, a CPU variant would be generated on a GPU host and run with `device="cpu"` (since `@skip_if_lt_x_gpu(1)` only checks accelerator availability, not whether the current variant is an accelerator), diverging from the original single-on-accelerator run. `except_for=("cpu",)` keeps the original behavior by excluding the CPU variant at instantiation time.

## Test Plan
Verified on an 8-device accelerator backend with the threaded reference PG (the `HardwareClassification` infrastructure from #186918 are pre-applied on the verify host):

```bash
python test/distributed/test_multi_threaded_pg.py
# Ran 22 tests in 1.014s — OK

python test/distributed/test_multi_threaded_pg.py --hw-classification GENERIC ACCELERATOR
# total=22, passed=22, unclassified=0, mismatched=0

python test/distributed/test_multi_threaded_pg.py --hw-classification GENERIC
# total=22, passed=21, mismatched=1 (the ACCELERATOR class excluded, as expected)
```

## Notes
- This is part of the test case refactoring initiative tracked in #185590.

### CODE REVIEW

**Comment by twthighd** [comment]
@wjlFlyer @pjyFight @FuDdd @fffrog @lyriexs666 @jishuangfeng @JamesD18 @JiasenTian

**Comment by jishuangfeng** [comment]
The class split, GENERIC/ACCELERATOR labeling, and instantiate_device_type_tests parameterization all look correct. The @skip_if_lt_x_gpu(1) retention on test_bwd_sees_fwd_pg is appropriate given that #187650 makes that decorator hardware-agnostic, and the dependency is clearly noted in the PR description.
LGTM.

**Review by jishuangfeng — APPROVED** [review]
LGTM

**Review by wjlFlyer — APPROVED** [review]

File: test/distributed/test_multi_threaded_pg.py [hunk 1]
```diff
@@ -335,11 +349,14 @@ def backward(ctx, grad_output):
...
-            [dist.get_rank()], dtype=torch.float, device=device_type, requires_grad=True
+            [dist.get_rank()], dtype=torch.float, device=device, requires_grad=True
         )
         x = MyFunc.apply(x)
         x.sum().backward()
 
 
+instantiate_device_type_tests(
+    TestCollectivesWithBaseClassDevice, globals(), except_for=("cpu",), allow_xpu=True
+)
```
Comment (fffrog) [inline test/distributed/test_multi_threaded_pg.py:358]:
```suggestion
instantiate_device_type_tests(
    TestCollectivesWithBaseClassDevice, globals(), allow_xpu=True
)
```

File: test/distributed/test_multi_threaded_pg.py [hunk 2]
```diff
@@ -335,11 +349,14 @@ def backward(ctx, grad_output):
...
         x = torch.tensor(
-            [dist.get_rank()], dtype=torch.float, device=device_type, requires_grad=True
+            [dist.get_rank()], dtype=torch.float, device=device, requires_grad=True
         )
         x = MyFunc.apply(x)
         x.sum().backward()
 
 
+instantiate_device_type_tests(
+    TestCollectivesWithBaseClassDevice, globals(), except_for=("cpu",), allow_xpu=True
```
Comment (fffrog) [inline test/distributed/test_multi_threaded_pg.py:357]: Because the device_type could be `CPU`, right?

**Review by fffrog — CHANGES_REQUESTED** [review]
SGTM, thank you for contributing this one.

File: test/distributed/test_multi_threaded_pg.py (hunk 1 above)
Comment (twthighd) [reply test/distributed/test_multi_threaded_pg.py:358]: Thanks for the review!  Adopted in `2de78742` — dropped `except_for=("cpu",)`.

File: test/distributed/test_multi_threaded_pg.py (hunk 2 above)
Comment (twthighd) [reply test/distributed/test_multi_threaded_pg.py:357]: Yes, exactly. With `except_for` removed, the CPU variant is generated, and
`device` is injected as `"cpu"` for it. `test_bwd_sees_fwd_pg` keeps
`@skip_if_lt_x_gpu(1)` so it is skipped on the CPU variant (no accelerator
available) while running on-device for accelerator variants. So the
`@skip_if_lt_x_gpu(1)` is no longer dead — it now genuinely gates the
accelerator-only method against the CPU variant, and leaves
the class open to future CPU-runnable `test_*` methods without re-editing the
instantiation call.

File: test/distributed/test_multi_threaded_pg.py [hunk 3]
```diff
@@ -335,11 +349,14 @@ def backward(ctx, grad_output):
                 return grad_output * result
 
         x = torch.tensor(
-            [dist.get_rank()], dtype=torch.float, device=device_type, requires_grad=True
+            [dist.get_rank()], dtype=torch.float, device=device, requires_grad=True
```
Comment (wjlFlyer) [inline test/distributed/test_multi_threaded_pg.py:350]: There is a quesiton. If you remove `except_for=("cpu",)` in `instantiate_device_type_tests` method, class `TestCollectivesWithBaseClassDeviceCPU` will be instantiated.

Will `TestCollectivesWithBaseClassDeviceCPU`  be executed on gpu server? If so, the parameter `device` would be `cpu`. But for original logic, if this method is executed on gpu server, the `device_type` would be `gpu` because it is generated by `device_type = acc.type if (acc := torch.accelerator.current_accelerator()) else "cpu"`.

Comment (twthighd) [reply test/distributed/test_multi_threaded_pg.py:350]: Thanks for catching this — it's a valid concern. With `except_for` removed, the CPU variant runs on a GPU host with `device="cpu"` (since `@skip_if_lt_x_gpu(1)` only checks accelerator availability, not the current variant), diverging from the original single-GPU run.

Fix: restore `except_for=("cpu",)` (no CPU variant generated) and drop the now-redundant `@skip_if_lt_x_gpu(1)` (pure-CPU skip is now handled at instantiation).

This keeps original behavior + decoupling for other accelerators.

**Review by wjlFlyer — APPROVED** [review]

**Review by KarhouTam — APPROVED** [review]
Thanks.
cc @fffrog

**Review by KarhouTam — CHANGES_REQUESTED** [review]
Please fix conflict

**Comment by twthighd** [comment]
> Please fix conflict

Conflict resolved  — rebased onto latest main . Notes on how the upstream changes were folded in:

test_collectives_issued_from_side_stream (#195281) read the module-level device_type constant this PR deletes, so it moved into TestCollectivesWithBaseClassDevice as well: device parameter, helpers derived from self.device_type, and its @skip_if_lt_x_gpu(1) dropped (redundant with except_for=("cpu",), same reasoning as test_bwd_sees_fwd_pg). The method body is upstream verbatim apart from the signature/decorator.
TestThreadLocalWorld (#196786) is labeled GENERIC so every collected class in the file carries a hw_classification.

**Review by KarhouTam — APPROVED** [review]
Thanks.
cc @fffrog
